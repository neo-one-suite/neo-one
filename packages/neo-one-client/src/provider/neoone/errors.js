/* @flow */
import { CustomError } from '@neo-one/utils';

export type JSONRPCErrorResponse = {|
  code: number,
  message: string,
  data?: $FlowFixMe,
|};

export class JSONRPCError extends CustomError {
  responseError: JSONRPCErrorResponse;
  code: string;

  constructor(responseError: JSONRPCErrorResponse) {
    super(responseError.message);
    this.responseError = responseError;
    this.code = 'JSON_RPC';
  }
}

export class InvalidRPCResponseError extends CustomError {
  code: string;
  constructor() {
    super('Did not receive valid rpc response');
    this.code = 'INVALID_RPC_RESPONSE';
  }
}

export class HTTPError extends CustomError {
  status: number;
  text: ?string;
  code: string;

  constructor(status: number, text: ?string) {
    let message = `HTTP Error ${status}`;
    if (text != null) {
      message = `${message}: ${text}`;
    }
    super(message);
    this.status = status;
    this.text = text;
    this.code = 'HTTP';
  }
}

export class MissingTransactionDataError extends CustomError {
  code: string = 'MISSING_TRANSACTION_DATA';

  constructor(hash: string) {
    super(`Missing transaction data for transaction ${hash}`);
  }
}
