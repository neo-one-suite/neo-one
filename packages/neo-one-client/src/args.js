/* @flow */
import BigNumber from 'bignumber.js';

import _ from 'lodash';
import { common } from '@neo-one/client-core';

import type {
  ABI,
  ABIEvent,
  ABIFunction,
  ABIParameter,
  ABIReturn,
  AddressString,
  AttributeArg,
  BlockFilter,
  BufferString,
  GetOptions,
  Hash160String,
  Hash256String,
  PublicKeyString,
  TransactionOptions,
} from './types';
import { InvalidArgumentError, InvalidNamedArgumentError } from './errors';

import { addressToScriptHash } from './helpers';
import converters from './user/converters';

export const assertString = (name: string, param: mixed): string => {
  if (param == null || typeof param !== 'string') {
    throw new InvalidArgumentError(
      `Expected string for ${name}, found: ${String(param)}`,
    );
  }

  return param;
};

export const assertNullableString = (name: string, param?: mixed): ?string => {
  if (param == null) {
    return param;
  }

  return assertString(name, param);
};

export const assertAddress = (address: mixed): AddressString => {
  if (address == null || typeof address !== 'string') {
    throw new InvalidArgumentError(
      `Address argument was not a string: ${String(address)}`,
    );
  }

  try {
    addressToScriptHash(address);
  } catch (error) {
    throw new InvalidArgumentError(`Invalid address: ${address}`);
  }

  return address;
};

export const assertHash160 = (hash: mixed): Hash160String => {
  const value = assertString('Hash160', hash);

  if (!value.startsWith('0x')) {
    throw new InvalidArgumentError(
      `Hash160 must start with '0x', found: ${String(hash)}`,
    );
  }

  try {
    common.stringToUInt160(value);
  } catch (error) {
    throw new InvalidArgumentError(
      `Invalid Hash160 param, found: ${String(hash)}`,
    );
  }

  return value;
};

export const assertHash256 = (hash: mixed): Hash256String => {
  const value = assertString('Hash256', hash);

  if (!value.startsWith('0x')) {
    throw new InvalidArgumentError(
      `Hash256 must start with '0x', found: ${String(hash)}`,
    );
  }

  try {
    common.stringToUInt256(value);
  } catch (error) {
    throw new InvalidArgumentError(
      `Invalid Hash256 param, found: ${String(hash)}`,
    );
  }

  return value;
};

export const assertBuffer = (buffer: mixed): BufferString => {
  const value = assertString('Buffer', buffer);
  if (Buffer.from(value, 'hex').toString('hex') !== value.toLowerCase()) {
    throw new InvalidArgumentError(
      `Expected hex string, found: ${String(buffer)}`,
    );
  }

  return value;
};

export const assertPublicKey = (publicKey: mixed): PublicKeyString => {
  const value = assertBuffer(publicKey);
  try {
    common.stringToECPoint(value);
    return value;
  } catch (error) {
    throw new InvalidArgumentError(
      `Expected PublicKey, found: ${String(publicKey)}`,
    );
  }
};

export const assertBigNumber = (value: mixed): BigNumber => {
  if (value == null || !BigNumber.isBigNumber(value)) {
    throw new InvalidArgumentError(
      `Expected BigNumber, found: ${String(value)}`,
    );
  }

  return (value: $FlowFixMe);
};

export const assertNullableBigNumber = (value: mixed): ?BigNumber => {
  if (value == null) {
    return value;
  }

  return assertBigNumber(value);
};

export const assertBoolean = (value: mixed): boolean => {
  if (value == null || typeof value !== 'boolean') {
    throw new InvalidArgumentError(`Expected boolean, found: ${String(value)}`);
  }

  return value;
};

export const assertNumber = (value: mixed): number => {
  if (value == null || typeof value !== 'number') {
    throw new InvalidArgumentError(`Expected number, found: ${String(value)}`);
  }

  return value;
};

export const assertNullableNumber = (value: mixed): ?number => {
  if (value != null && typeof value !== 'number') {
    throw new InvalidArgumentError(`Expected number, found: ${String(value)}`);
  }

  return value;
};

export const assertArray = (value: mixed): Array<mixed> => {
  if (!Array.isArray(value)) {
    throw new InvalidArgumentError(`Expected Array, found: ${String(value)}`);
  }

  return value;
};

export const assertBlockFilter = (filter: mixed): ?BlockFilter => {
  if (filter == null) {
    return filter;
  }

  if (typeof filter !== 'object') {
    throw new InvalidArgumentError(
      `Invalid BlockFilter param, found: ${String(filter)}`,
    );
  }

  if (_.isEmpty(filter)) {
    return ({}: $FlowFixMe);
  }

  if (filter.indexStart != null && typeof filter.indexStart !== 'number') {
    throw new InvalidArgumentError(
      `Invalid BlockFilter param, found: ${String(filter)}`,
    );
  }

  if (filter.indexStop != null && typeof filter.indexStop !== 'number') {
    throw new InvalidArgumentError(
      `Invalid BlockFilter param, found: ${String(filter)}`,
    );
  }

  if (
    filter.indexStart != null &&
    filter.indexStop != null &&
    filter.indexStart > filter.indexStop
  ) {
    throw new InvalidArgumentError(
      `Invalid BlockFilter param, found: ${String(filter)}`,
    );
  }

  return (filter: $FlowFixMe);
};

export const assertGetOptions = (options: mixed): ?GetOptions => {
  if (options == null) {
    return options;
  }

  if (typeof options !== 'object') {
    throw new InvalidArgumentError(
      `Invalid GetOptions param, found: ${String(options)}`,
    );
  }

  if (_.isEmpty(options)) {
    return ({}: $FlowFixMe);
  }

  if (options.timeout != null && typeof options.timeout !== 'number') {
    throw new InvalidArgumentError(
      `Invalid GetOptions param, found: ${String(options)}`,
    );
  }

  return (options: $FlowFixMe);
};

const ABI_TYPES = new Set([
  'Signature',
  'Boolean',
  'Hash160',
  'Hash256',
  'ByteArray',
  'PublicKey',
  'String',
  'InteropInterface',
  'Void',
  'Integer',
  'Array',
]);

export const assertABIReturn = (value: mixed): ABIReturn => {
  if (value == null || typeof value !== 'object') {
    throw new InvalidArgumentError(
      `Invalid ABI return, found: ${String(value)}`,
    );
  }

  if (!ABI_TYPES.has(value.type)) {
    throw new InvalidArgumentError(
      `Invalid ABI return, found: ${String(value)}`,
    );
  }

  if (value.type === 'Array') {
    assertABIReturn(value.value);
  } else if (value.type === 'Integer') {
    assertNumber(value.decimals);
  }

  return (value: $FlowFixMe);
};

export const assertABIParameter = (value: mixed): ABIParameter => {
  if (value == null || typeof value !== 'object') {
    throw new InvalidArgumentError(
      `Invalid ABI parameter, found: ${String(value)}`,
    );
  }

  assertString('name', value.name);
  assertABIReturn(value);
  return (value: $FlowFixMe);
};

export const assertABIFunction = (value: mixed): ABIFunction => {
  if (value == null || typeof value !== 'object') {
    throw new InvalidArgumentError(
      `Invalid ABI function, found: ${String(value)}`,
    );
  }

  assertString('name', value.name);
  if (value.constant != null) {
    assertBoolean(value.constant);
  }

  if (value.parameters != null) {
    assertArray(value.parameters).map((parameter) =>
      assertABIParameter(parameter),
    );
  }

  assertABIReturn(value.returnType);

  return (value: $FlowFixMe);
};

export const assertABIEvent = (value: mixed): ABIEvent => {
  if (value == null || typeof value !== 'object') {
    throw new InvalidArgumentError(
      `Invalid ABI event, found: ${String(value)}`,
    );
  }

  assertString('name', value.name);
  assertArray(value.parameters).map((parameter) =>
    assertABIParameter(parameter),
  );

  return (value: $FlowFixMe);
};

export const assertABI = (abi: mixed): ABI => {
  if (abi == null || typeof abi !== 'object') {
    throw new InvalidArgumentError(`Invalid ABI param, found: ${String(abi)}`);
  }

  assertArray(abi.functions).map((value) => assertABIFunction(value));
  if (abi.events != null) {
    assertArray(abi.events).map((event) => assertABIEvent(event));
  }

  return (abi: $FlowFixMe);
};

export const assertAttributeArg = (attribute?: mixed): AttributeArg => {
  try {
    converters.attribute((attribute: $FlowFixMe));
    return (attribute: $FlowFixMe);
  } catch (error) {
    throw new InvalidNamedArgumentError('AttributeArg', attribute);
  }
};

export const assertTransactionOptions = (
  options?: mixed,
): ?TransactionOptions => {
  if (options == null) {
    return options;
  }

  if (typeof options !== 'object') {
    throw new InvalidNamedArgumentError('TransactionOptions', options);
  }

  const { from } = options;
  if (from != null) {
    if (typeof from !== 'object') {
      throw new InvalidNamedArgumentError('TransactionOptions', options);
    }
    assertString('TransactionOptions', from.network);
    assertString('TransactionOptions', from.address);
  }

  if (options.attributes != null) {
    if (!Array.isArray(options.attributes)) {
      throw new InvalidNamedArgumentError('TransactionOptions', options);
    }
    options.attributes.forEach((attribute) => {
      assertAttributeArg(attribute);
    });
  }

  assertNullableBigNumber(options.networkFee);

  return (options: $FlowFixMe);
};
