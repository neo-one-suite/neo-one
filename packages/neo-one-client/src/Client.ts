import { ABI, ScriptBuilderParam } from '@neo-one/client-core';
import BigNumber from 'bignumber.js';
import * as args from './args';
import { ClientBase } from './ClientBase';
import { ReadClient } from './ReadClient';
import { createSmartContract } from './sc';
import {
  AddressString,
  AssetRegister,
  ContractRegister,
  Hash256String,
  InvokeTransactionOptions,
  NetworkType,
  Param,
  PublishReceipt,
  RawCallReceipt,
  RawInvokeReceipt,
  RegisterAssetReceipt,
  SmartContract,
  SmartContractAny,
  SmartContractDefinition,
  SourceMaps,
  TransactionOptions,
  TransactionResult,
  Transfer,
  UserAccountProvider,
} from './types';
const mutableClients: Client[] = [];

export class Client<
  // tslint:disable-next-line no-any
  TUserAccountProviders extends { readonly [K: string]: UserAccountProvider } = any
> extends ClientBase<TUserAccountProviders> {
  public static inject(provider: UserAccountProvider): void {
    mutableClients.forEach((client) => client.inject(provider));
  }

  public constructor(providersIn: TUserAccountProviders) {
    super(providersIn);
    mutableClients.push(this);
  }

  public async transfer(
    amount: BigNumber,
    asset: Hash256String,
    to: AddressString,
    options?: TransactionOptions,
  ): Promise<TransactionResult>;
  public async transfer(transfers: ReadonlyArray<Transfer>, options?: TransactionOptions): Promise<TransactionResult>;
  // tslint:disable-next-line readonly-array no-any
  public async transfer(...argsIn: any[]): Promise<TransactionResult> {
    const { transfers, options } = this.getTransfersOptions(argsIn);

    return this.getProvider(options).transfer(transfers, options);
  }

  public async claim(options?: TransactionOptions): Promise<TransactionResult> {
    return this.getProvider(options).claim(args.assertTransactionOptions('options', options));
  }

  public async publish(
    contract: ContractRegister,
    options?: TransactionOptions,
  ): Promise<TransactionResult<PublishReceipt>> {
    return this.getProvider(options).publish(
      args.assertContractRegister('contract', contract),
      args.assertTransactionOptions('options', options),
    );
  }

  public async publishAndDeploy(
    contract: ContractRegister,
    abi: ABI,
    params: ReadonlyArray<Param> = [],
    options?: TransactionOptions,
    sourceMaps: SourceMaps = {},
  ): Promise<TransactionResult<PublishReceipt>> {
    return this.getProvider(options).publishAndDeploy(
      args.assertContractRegister('contract', contract),
      args.assertABI('abi', abi),
      params,
      args.assertTransactionOptions('options', options),
      sourceMaps,
    );
  }

  public async registerAsset(
    asset: AssetRegister,
    options?: TransactionOptions,
  ): Promise<TransactionResult<RegisterAssetReceipt>> {
    return this.getProvider(options).registerAsset(
      args.assertAssetRegister('asset', asset),
      args.assertTransactionOptions('options', options),
    );
  }

  public async issue(
    amount: BigNumber,
    asset: Hash256String,
    to: AddressString,
    options?: TransactionOptions,
  ): Promise<TransactionResult>;
  public async issue(transfers: ReadonlyArray<Transfer>, options?: TransactionOptions): Promise<TransactionResult>;
  // tslint:disable-next-line readonly-array no-any
  public async issue(...argsIn: any[]): Promise<TransactionResult> {
    const { transfers, options } = this.getTransfersOptions(argsIn);

    return this.getProvider(options).issue(transfers, options);
  }

  public read(network: NetworkType): ReadClient {
    return new ReadClient(
      this.getNetworkProvider(args.assertString('network', network)).read(args.assertString('network', network)),
    );
  }

  // tslint:disable-next-line no-any
  public smartContract<T extends SmartContract<any> = SmartContractAny>(definition: SmartContractDefinition): T {
    return createSmartContract({
      definition: args.assertSmartContractDefinition('definition', definition),
      client: this,
      // tslint:disable-next-line no-any
    }) as any;
  }

  public inject<K>(
    provider: K extends keyof TUserAccountProviders ? TUserAccountProviders[K] : UserAccountProvider,
  ): void {
    this.providers$.next(
      // tslint:disable-next-line prefer-object-spread
      Object.assign({}, this.providers, {
        [provider.type]: provider,
      }),
    );

    this.selectedProvider$.next(provider);
  }

  // internal
  public async __invoke(
    contract: AddressString,
    method: string,
    params: ReadonlyArray<ScriptBuilderParam | undefined>,
    paramsZipped: ReadonlyArray<[string, Param | undefined]>,
    verify: boolean,
    options?: InvokeTransactionOptions,
    sourceMaps: SourceMaps = {},
  ): Promise<TransactionResult<RawInvokeReceipt>> {
    return this.getProvider(options).invoke(contract, method, params, paramsZipped, verify, options, sourceMaps);
  }

  public async __call(
    contract: AddressString,
    method: string,
    params: ReadonlyArray<ScriptBuilderParam | undefined>,
    options?: TransactionOptions,
  ): Promise<RawCallReceipt> {
    return this.getProvider(options).call(contract, method, params, options);
  }

  private getTransfersOptions(
    // tslint:disable-next-line no-any
    argsIn: ReadonlyArray<any>,
  ): {
    readonly transfers: ReadonlyArray<Transfer>;
    readonly options?: TransactionOptions;
  } {
    let transfers;
    let options;
    if (argsIn.length >= 3) {
      transfers = [
        {
          amount: argsIn[0],
          asset: argsIn[1],
          to: argsIn[2],
        },
      ];

      options = argsIn[3];
    } else {
      transfers = argsIn[0];
      options = argsIn[1];
    }

    return {
      transfers: args.assertTransfers('transfers', transfers),
      options: args.assertTransactionOptions('options', options),
    };
  }
}
