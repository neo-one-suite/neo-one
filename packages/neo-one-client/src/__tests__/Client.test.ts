import BigNumber from 'bignumber.js';
import { of } from 'rxjs';
import { take } from 'rxjs/operators';
import * as argAssertions from '../args';
import { Client } from '../Client';
import { InvalidArgumentError, UnknownNetworkError } from '../errors';
import { createSmartContract } from '../sc/createSmartContract';
import { UserAccountProvider } from '../types';

jest.mock('../sc/createSmartContract');

describe('Client', () => {
  const id1 = { network: 'net1', address: 'APyEx5f4Zm4oCHwFWiSTaph1fPBxZacYVR' };
  const id2 = { network: 'net2', address: 'Aew1QakLBrMqn9pmYpVrgwKPXCqTfzd6sZ' };
  const account1 = {
    type: 'test1',
    id: id1,
    name: 'account1',
    scriptHash: 'hash160-1',
    publicKey: 'pukey-1',
  };

  const account2 = {
    type: 'test2',
    id: id2,
    name: 'account2',
    scriptHash: 'hash160-2',
    publicKey: 'pubkey-2',
  };

  const network1 = 'net1';
  const network2 = 'net2';
  const provider1: UserAccountProvider = {
    type: 'test1',
    currentAccount$: of(account1),
    accounts$: of([account1]),
    networks$: of(['net1']),
    getCurrentAccount: () => account1,
    getAccounts: () => [account1],
    getNetworks: () => [network1],
    deleteAccount: () => [],
    updateAccountName: () => [],
    read: () => [],
  } as any;

  const provider2: UserAccountProvider = {
    type: 'test2',
    currentAccount$: of(account2),
    accounts$: of([account2]),
    networks$: of(['net2']),
    getCurrentAccount: () => account2,
    getAccounts: () => [account2],
    getNetworks: () => [network2],
    deleteAccount: () => [],
    updateAccountName: () => [],
    read: () => [],
  } as any;

  let client = new Client({ test1: provider1, test2: provider2 } as any);
  beforeEach(() => {
    client = new Client({ test1: provider1, test2: provider2 } as any);
  });

  test('Client constructor throws error on undefined client', () => {
    function testError() {
      client = new Client({});
    }

    expect(testError).toThrow(new Error('At least one provider is required') as any);
  });

  test('Client constructor throws error on mismatched provider type', () => {
    function testError() {
      client = new Client({
        test: {
          type: 'invalid',
          getCurrentAccount: () => {
            // do nothing
          },
        } as any,
      });
    }

    expect(testError).toThrow(new Error('Provider keys must be named the same as their type') as any);
  });

  test('Client constructor sets up currentAcount$ observable', async () => {
    const result = await client.currentAccount$.pipe(take(1)).toPromise();
    expect(result).toEqual(account1);
  });

  test('Client constructor sets up accounts$ observable', async () => {
    const result = await client.accounts$.pipe(take(1)).toPromise();
    expect(result).toEqual([account1, account2]);
  });

  test('Client constructor sets up networks$ observable', async () => {
    const result = await client.networks$.pipe(take(1)).toPromise();
    expect(result).toEqual(['net1', 'net2']);
  });

  test('get providers', async () => {
    const result = client.providers;
    expect(result).toEqual({
      test1: provider1,
      test2: provider2,
    });
  });

  test('selectAccount throws error on unknown account', async () => {
    const accountID = { network: 'fakeNet', address: 'fakeAddr' };

    const result = client.selectAccount(accountID);

    await expect(result).rejects.toEqual(
      new InvalidArgumentError(`Invalid argument for UserAccountID: Error: Invalid address: ${accountID.address}`),
    );
  });

  test('selectAccount', async () => {
    (provider2 as any).selectAccount = jest.fn(() => Promise.resolve());

    expect(client.getCurrentAccount()).toEqual(account1);
    await client.selectAccount(id2);
    expect(client.getCurrentAccount()).toEqual(account2);
  });

  test('getAccounts', () => {
    const expected = [account1, account2];

    const result = client.getAccounts();
    expect(result).toEqual(expected);
  });

  test('getAccount', () => {
    const expected = account1;

    const result = client.getAccount(id1);
    expect(result).toEqual(expected);
  });

  test('getAccount throws error on nonexistant account', () => {
    const idFake = { network: 'fakeNet', address: 'fakeAddr' };

    function testError() {
      return client.getAccount(idFake);
    }
    expect(testError).toThrow(new InvalidArgumentError(
      `Invalid argument for UserAccountID: Error: Invalid address: ${idFake.address}`,
    ) as any);
  });

  test('deleteAccount', async () => {
    const mocked = jest.fn(() => Promise.resolve());
    (provider2 as any).deleteAccount = mocked;

    await client.deleteAccount(id2);
    expect(mocked.mock.calls).toMatchSnapshot();
  });

  test('updateAccountName', async () => {
    const mocked = jest.fn(() => Promise.resolve());
    (provider2 as any).updateAccountName = mocked;

    await client.updateAccountName({ id: id2, name: 'newName' });
    expect(mocked.mock.calls).toMatchSnapshot();
  });

  test('getNetworks', () => {
    const expected = [network1, network2];

    const result = client.getNetworks();
    expect(result).toEqual(expected);
  });

  test('read', () => {
    const mocked = jest.fn(() => {
      // do nothing
    });
    (provider1 as any).read = mocked;

    const result = client.read('net1');
    expect(result).toMatchSnapshot();
    expect(mocked.mock.calls).toMatchSnapshot();
  });

  test('read throws error on invalid network', () => {
    function testError() {
      return client.read('fakeNet');
    }
    expect(testError).toThrow(new UnknownNetworkError('fakeNet') as any);
  });

  test('inject', () => {
    client = new Client({ test1: provider1 } as any);

    expect(client.getCurrentAccount()).toEqual(account1);
    expect(client.providers).toEqual({ test1: provider1 });
    // any
    client.inject(provider2);

    expect(client.getCurrentAccount()).toEqual(account2);
    expect(client.providers).toEqual({
      test1: provider1,
      test2: provider2,
    });
  });

  test('static inject', () => {
    client = new Client({ test1: provider1 } as any);

    expect(client.getCurrentAccount()).toEqual(account1);
    expect(client.providers).toEqual({ test1: provider1 });
    // any
    Client.inject(provider2);

    expect(client.getCurrentAccount()).toEqual(account2);
    expect(client.providers).toEqual({
      test1: provider1,
      test2: provider2,
    });
  });

  const testCases = [
    {
      method: 'getCurrentAccount',
      args: [],
    },

    {
      method: 'transfer',
      args: [
        [
          {
            amount: new BigNumber(10),
            asset: '0x7f48028c38117ac9e42c8e1f6f06ae027cdbb904eaf1a0bdc30c9d81694e045c',
            to: 'APyEx5f4Zm4oCHwFWiSTaph1fPBxZacYVR',
          },
        ],
        undefined,
      ],
    },

    {
      method: 'transfer',
      args: [
        [
          {
            amount: new BigNumber(10),
            asset: '0x7f48028c38117ac9e42c8e1f6f06ae027cdbb904eaf1a0bdc30c9d81694e045c',
            to: 'APyEx5f4Zm4oCHwFWiSTaph1fPBxZacYVR',
          },
        ],
      ],
    },

    {
      method: 'claim',
      args: [],
    },

    {
      method: 'publish',
      args: [
        {
          script: '02028a99826e',
          parameters: ['Signature'],
          returnType: 'Signature',
          name: 'contract',
          codeVersion: 'codeVersion',
          author: 'author',
          email: 'email',
          description: 'description',
          properties: {
            storage: true,
            dynamicInvoke: true,
            payable: true,
          },
        },
      ],
    },

    {
      method: 'registerAsset',
      args: [
        {
          assetType: 'CreditFlag',
          name: 'asset',
          amount: new BigNumber(10),
          precision: 20,
          owner: '02028a99826edc0c97d18e22b6932373d908d323aa7f92656a77ec26e8861699ef',
          admin: 'APyEx5f4Zm4oCHwFWiSTaph1fPBxZacYVR',
          issuer: 'APyEx5f4Zm4oCHwFWiSTaph1fPBxZacYVR',
        },
      ],
    },

    {
      method: 'issue',
      args: [
        [
          {
            amount: new BigNumber(10),
            asset: '0x7f48028c38117ac9e42c8e1f6f06ae027cdbb904eaf1a0bdc30c9d81694e045c',
            to: 'APyEx5f4Zm4oCHwFWiSTaph1fPBxZacYVR',
          },
        ],
        undefined,
      ],
    },

    {
      method: 'invoke',
      args: ['0xecc6b20d3ccac1ee9ef109af5a7cdb85706b1df9', 1, 2, 3, true],
      providerMethod: 'invoke',
    },

    {
      method: 'call',
      args: ['0xecc6b20d3ccac1ee9ef109af5a7cdb85706b1df9', 1, 2],
      providerMethod: 'call',
    },
  ];

  for (const testCase of testCases) {
    const { method, args } = testCase;
    const expected = '10';

    let providerMethod = method;
    if (testCase.providerMethod != undefined) {
      providerMethod = testCase.providerMethod;
    }
    test(method, async () => {
      (provider1 as any)[providerMethod] = jest.fn(() => Promise.resolve(expected));

      if (method === 'claim' || method === 'publish') {
        (argAssertions as any).assertTransactionOptions = jest.fn(() => true);
      }
      const result = await (client as any)[method](...args);
      expect(result).toEqual(expected);
    });
  }

  test('smartContract', () => {
    const expected = {
      networks: {
        main: {
          hash: '0xecc6b20d3ccac1ee9ef109af5a7cdb85706b1df9',
        },
      },
      abi: {
        functions: [],
      },
    };
    (createSmartContract as any).mockImplementation(() => expected);
    const result = client.smartContract(expected as any);
    expect(result).toEqual(expected);
  });
});
