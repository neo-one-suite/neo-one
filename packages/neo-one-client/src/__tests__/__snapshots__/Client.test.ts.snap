// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client __call 1`] = `
Array [
  Array [
    "ALq7AWrhAueN6mJNqk6FHJjnsEoPRytLdW",
    "deploy",
    Array [],
    undefined,
  ],
]
`;

exports[`Client __invoke 1`] = `
Array [
  Array [
    "ALq7AWrhAueN6mJNqk6FHJjnsEoPRytLdW",
    "deploy",
    Array [],
    Array [],
    true,
    undefined,
    Object {},
  ],
]
`;

exports[`Client claim 1`] = `Array []`;

exports[`Client constructor automatically selects an account 1`] = `
Array [
  Array [
    Object {
      "address": "ALfnhLg7rUyL6Jr98bzzoxz5J7m64fbR4s",
      "network": "main",
    },
  ],
]
`;

exports[`Client constructor does not automatically select an account if one exists at the time of selection 1`] = `Array []`;

exports[`Client constructor throws on invalid provider key 1`] = `"Provider keys must be named the same as their type"`;

exports[`Client constructor throws on no provider 1`] = `"At least one provider is required"`;

exports[`Client deleteAccount 1`] = `
Array [
  Array [
    Object {
      "address": "ALfnhLg7rUyL6Jr98bzzoxz5J7m64fbR4s",
      "network": "main",
    },
  ],
]
`;

exports[`Client getAccount - throws on unknown account 1`] = `"Unknown account ALq7AWrhAueN6mJNqk6FHJjnsEoPRytLdW"`;

exports[`Client issue - array 1`] = `
Array [
  Array [
    Array [
      Object {
        "amount": "10",
        "asset": "0xc56f33fc6ecfcd0c225c4ab356fee59390af8560be0e930faebe74a6daff7c9b",
        "to": "ALq7AWrhAueN6mJNqk6FHJjnsEoPRytLdW",
      },
    ],
    Object {
      "attributes": Array [],
      "from": Object {
        "address": "AVf4UGKevVrMR1j3UkPsuoYKSC4ocoAkKx",
        "network": "main",
      },
      "monitor": undefined,
      "networkFee": undefined,
    },
  ],
]
`;

exports[`Client issue - simple 1`] = `
Array [
  Array [
    Array [
      Object {
        "amount": "10",
        "asset": "0xc56f33fc6ecfcd0c225c4ab356fee59390af8560be0e930faebe74a6daff7c9b",
        "to": "ALq7AWrhAueN6mJNqk6FHJjnsEoPRytLdW",
      },
    ],
    undefined,
  ],
]
`;

exports[`Client publish 1`] = `
Array [
  Array [
    Object {
      "author": "dicarlo2",
      "codeVersion": "1.0",
      "description": "Hello World",
      "dynamicInvoke": false,
      "email": "alex.dicarlo@neotracker.io",
      "name": "MyContract",
      "parameters": Array [
        "Address",
        "Buffer",
      ],
      "payable": true,
      "returnType": "Buffer",
      "script": "5f8d70",
      "storage": true,
    },
    undefined,
  ],
]
`;

exports[`Client publishAndDeploy 1`] = `
Array [
  Array [
    Object {
      "author": "dicarlo2",
      "codeVersion": "1.0",
      "description": "Hello World",
      "dynamicInvoke": false,
      "email": "alex.dicarlo@neotracker.io",
      "name": "MyContract",
      "parameters": Array [
        "Address",
        "Buffer",
      ],
      "payable": true,
      "returnType": "Buffer",
      "script": "5f8d70",
      "storage": true,
    },
    Object {
      "events": Array [],
      "functions": Array [
        Object {
          "constant": undefined,
          "name": "deploy",
          "parameters": Array [],
          "returnType": Object {
            "optional": undefined,
            "type": "Boolean",
          },
          "verify": undefined,
        },
      ],
    },
    Array [],
    undefined,
    Object {},
  ],
]
`;

exports[`Client read - throws on unknown network 1`] = `"Unknown network unknown"`;

exports[`Client read 1`] = `
Array [
  Array [
    "main",
  ],
]
`;

exports[`Client read 2`] = `
ReadClient {
  "dataProvider": Object {},
}
`;

exports[`Client registerAsset 1`] = `
Array [
  Array [
    Object {
      "admin": "ALq7AWrhAueN6mJNqk6FHJjnsEoPRytLdW",
      "amount": "100000000",
      "issuer": "ALfnhLg7rUyL6Jr98bzzoxz5J7m64fbR4s",
      "name": "TheToken",
      "owner": "02028a99826edc0c97d18e22b6932373d908d323aa7f92656a77ec26e8861699ef",
      "precision": 8,
      "type": "Token",
    },
    undefined,
  ],
]
`;

exports[`Client selectAccount 1`] = `Array []`;

exports[`Client smartContract 1`] = `
Object {
  "balanceOf": [Function],
  "decimals": [Function],
  "definition": Object {
    "abi": Object {
      "events": Array [
        Object {
          "name": "transfer",
          "parameters": Array [
            Object {
              "name": "from",
              "optional": undefined,
              "type": "Address",
            },
            Object {
              "name": "to",
              "optional": undefined,
              "type": "Address",
            },
            Object {
              "decimals": 8,
              "name": "amount",
              "optional": undefined,
              "type": "Integer",
            },
          ],
        },
      ],
      "functions": Array [
        Object {
          "constant": true,
          "name": "name",
          "parameters": Array [],
          "returnType": Object {
            "optional": undefined,
            "type": "String",
          },
          "verify": undefined,
        },
        Object {
          "constant": true,
          "name": "symbol",
          "parameters": Array [],
          "returnType": Object {
            "optional": undefined,
            "type": "String",
          },
          "verify": undefined,
        },
        Object {
          "constant": true,
          "name": "decimals",
          "parameters": Array [],
          "returnType": Object {
            "decimals": 0,
            "optional": undefined,
            "type": "Integer",
          },
          "verify": undefined,
        },
        Object {
          "constant": true,
          "name": "totalSupply",
          "parameters": Array [],
          "returnType": Object {
            "decimals": 8,
            "optional": undefined,
            "type": "Integer",
          },
          "verify": undefined,
        },
        Object {
          "constant": undefined,
          "name": "transfer",
          "parameters": Array [
            Object {
              "name": "from",
              "optional": undefined,
              "type": "Address",
            },
            Object {
              "name": "to",
              "optional": undefined,
              "type": "Address",
            },
            Object {
              "decimals": 8,
              "name": "value",
              "optional": undefined,
              "type": "Integer",
            },
          ],
          "returnType": Object {
            "optional": undefined,
            "type": "Boolean",
          },
          "verify": undefined,
        },
        Object {
          "constant": true,
          "name": "balanceOf",
          "parameters": Array [
            Object {
              "name": "account",
              "optional": undefined,
              "type": "Address",
            },
          ],
          "returnType": Object {
            "decimals": 8,
            "optional": undefined,
            "type": "Integer",
          },
          "verify": undefined,
        },
      ],
    },
    "networks": Object {
      "main": Object {
        "address": "ALq7AWrhAueN6mJNqk6FHJjnsEoPRytLdW",
      },
    },
    "sourceMaps": undefined,
  },
  "name": [Function],
  "read": [Function],
  "symbol": [Function],
  "totalSupply": [Function],
  "transfer": [Function],
}
`;

exports[`Client transfer - array 1`] = `
Array [
  Array [
    Array [
      Object {
        "amount": "10",
        "asset": "0xc56f33fc6ecfcd0c225c4ab356fee59390af8560be0e930faebe74a6daff7c9b",
        "to": "ALq7AWrhAueN6mJNqk6FHJjnsEoPRytLdW",
      },
    ],
    Object {
      "attributes": Array [],
      "from": Object {
        "address": "AVf4UGKevVrMR1j3UkPsuoYKSC4ocoAkKx",
        "network": "main",
      },
      "monitor": undefined,
      "networkFee": undefined,
    },
  ],
]
`;

exports[`Client transfer - simple 1`] = `
Array [
  Array [
    Array [
      Object {
        "amount": "10",
        "asset": "0xc56f33fc6ecfcd0c225c4ab356fee59390af8560be0e930faebe74a6daff7c9b",
        "to": "ALq7AWrhAueN6mJNqk6FHJjnsEoPRytLdW",
      },
    ],
    undefined,
  ],
]
`;

exports[`Client updateAccountName 1`] = `
Array [
  Array [
    Object {
      "id": Object {
        "address": "ALfnhLg7rUyL6Jr98bzzoxz5J7m64fbR4s",
        "network": "main",
      },
      "name": "newName",
    },
  ],
]
`;
